### ISSUE TITLE
vague discussion of speed with dplyr v. data.table

### ISSUE THREAD

**Issue by ljanda** [matloff/TidyverseSkeptic#12]
Your post often states that dplyr is slower than data.table and uses many qualifiers eg "much, much" slower but you do not clearly define what you mean by much slower or give clear information about when it is slower. You link out to a couple examples but this information should clearly be given within the post as this is a large part of your argument. 

You do pull some numbers from the h20ai site but your table is vague - it doesn't show what functions are being used or the size of the dataset or that it is in seconds and you don't consistently pull the smallest number 
<table border="1">
<tr>  <td>dplr</td>  <td>data.table</td>  </tr> 
<tr> <td>37.3</td>  <td>9.07</td> </tr>
<tr> <td>95.5</td>  <td>9.20</td> </tr>
<tr> <td>496</td>  <td>11.9</td> </tr>
</table>

Would redo in this way: 
Here are a couple examples of functions applied to multiple groups within dataset using the largest dataset tested in the link provided above. All outcomes are in seconds:

Dataset used: 1,000,000,000 rows x 9 columns (50GB)

function 1: 
```
# data.table
DT[, .v(v1=sum(v1)), by=id1]     
                                                                                                
# dpylr 
DF %>% 
  group_by(id1, .drop = TRUE) %>% 
  summarise(sum(v1))                                     
```
function 2: 
```
# data.table
 DT[, .v(v1=sum(v1)), by=.(id1, id2)]     
                                                                                 
# dplyr     
DF %>% 
  group_by(id1, id2, .drop = TRUE) %>% 
  summarise(sum(v1)) 
```
function 3: 
```
# data.table 
DT[, .v(v1=sum(v1)), v3 = mean(v3) by=id3]    
     
# dplyr                                       
DF %>% 
  group_by(id3 .drop = TRUE) %>% 
  summarise(sum(v1), mean(v3))    
```

<table border="1">
<tr> <td>groups</td>                                                   <td>dpylr</td>  <td>data.table</td>  </tr> 
<tr> <td>100 ad hoc groups of 10m rows, result 100 x 2</td>            <td>35.4</td>   <td>9.07</td>        </tr>
<tr> <td>10,000 ad hoc groups of 100,000 rows, result 10,000 x 3</td>  <td>97.5</td>   <td>9.20</td>        </tr>
<tr> <td>10m ad hoc groups of 100, rows, result 10m x 3</td>           <td>496</td>    <td>11.9</td>        </tr>
</table>

Note: The smaller datasets tested (10,000,000 rows x 9 columns, 0.5 GB and 100,000,000 rows x 9 columns, 5GB) also saw better speed with data.table but the difference was often a few seconds. 

I often use large-ish datasets (over a million rows, though often just a few million, with tens of variables) and have had very limited speed issues with dplyr. It is problematic to simply state that one should use data.table since it is faster without quantifying this claim. The data.table speed v. syntax tradeoff may not be worthwhile, especially for the many users who work with sizes where the speed difference is negligible. In some cases, data.table is an excellent option (I especially love fread()), but this post doesn't demonstrate the gains clearly or emphasize when it is useful. 

Finally, the graph you include following the table is fairly inscrutable (but looks dramatic since it is a graph quickly sloping up), would recommend rerunning it and making it with the seconds, rather than the confusing ratio. I went ahead and grabbed the code from the paper [you cited](https://github.com/WinVector/Examples/blob/master/dplyr/select_timing.Rmd) and edited it to make the suggested graph so here. I'd also like to point out that this example is a odd/niche since it is for selecting a variable out of an increasing number of variables, up to ~ 100,000 variables (always for 5 rows), and it is rare to work with such a large number of variables. The time for dplyr::select() only exceeds one second at 10,000 variables - something that would not matter in many cases. 

```
library("dplyr")
library("ggplot2")
library("rqdatatable")
library("cdata")
library(microbenchmark)
packageVersion("dplyr")
R.version

f <- function(k) {
  d <- rep(list(1:5), k)
  names(d) <- paste0("col_", seq_len(k))
  d <- data.frame(d)
  rownames(d) <- NULL
  d <- tbl_df(d)
  gc()
  tm <- microbenchmark::microbenchmark(
    select(d, col_1),
    d[, "col_1", drop = FALSE],
    times = 3L
  )
  td <- data.frame(tm)
  td$ncol <- ncol(d)
  td
}

times <- lapply(2^(0:17), f)
times <- data.frame(data.table::rbindlist(times))
times$seconds <- times$time/1e9

d1 <- 
  times %>% 
  mutate(b = ifelse(str_detect(expr, "select"), "dplyr", "base"), 
         b = factor(b))

ggplot(data = d1, aes(x = ncol, y = seconds, color = b)) +
  geom_point() + 
  geom_smooth(se = FALSE) + 
  scale_x_log10() + 
  scale_y_log10() +
  scale_color_brewer(palette = "Dark2") +
  scale_x_log10(labels = scales::comma) +
  labs(x = "Number of columns", y = "Seconds", 
       title = "Time to extract first column, dplyr::select() over base R [, ]") +
  theme_minimal() +
  theme(legend.position = "bottom", legend.title = element_blank()) 
```

![image](https://user-images.githubusercontent.com/17114370/61152792-57c4da80-a4b7-11e9-8233-133355751546.png)

**Comment by matloff** [comment]
I agree that the H2O presentation is incomplete. But there are various other analyses, and it is a generally known fact, that even RStudio concedes.

**Comment by ljanda** [comment]
The issue that you are hand-waving and posting unclear tables and graphs still holds. The time differences are often negligible with the dataset sizes many people work with. Also, R has challenges with "big data" regardless of packages used (which is why people turn to parallel computing packages, spark with R, etc). It is irresponsible to cherry-pick, obscure information, and post numbers and graphs without the relevant details.

**Comment by pmarchand1** [comment]
Does the comparison rely on the column to be in first position?

**Comment by ljanda** [comment]
@pmarchand1 yes the first example (that @matloff cited and I pulled the code from does), in the code it is "col_1", link: https://github.com/WinVector/Examples/blob/master/dplyr/select_timing.Rmd

**Comment by matloff** [comment]
See my earlier comment.
